monte_carlo.py

```diff
@@ -0,0 +1,195 @@
+                action = line.split(',')[4].strip()
+                self.policy[state] = action
+
+    def save(self, filename):
+        with open(filename, 'w') as f:
+            for state, action in self.policy.items():
+                f.write(f"{state[0]},{state[1]},{state[2]},{state[3]},{action}\n")
+    
+class MonteCarlo:
+
+    actions = {'increase_vx': (1, 0), 'decrease_vx': (-1, 0),
+           'increase_vy': (0, 1), 'decrease_vy': (0, -1),
+           'increase_vx_vy': (1, 1), 'decrease_vx_vy': (-1, -1),
+           'increase_vx_decrease_vy': (1, -1), 'decrease_vx_increase_vy': (-1, 1),
+           'no_change': (0, 0)}
+    max_velocity = 2
+    
+    def __init__(self, grid, gamma=0.9, num_episodes=10000, train_epsilon=0.9, test_epsilon=0.05, policy_filename=None, load_policy=False):
+        self.grid = grid
+
+        self.starts = np.where(self.grid == CELL.START)
+        self.starts = list(zip(self.starts[0], self.starts[1]))
+        self.goal = np.where(self.grid == CELL.TARGET)
+        self.goal = (self.goal[0][0], self.goal[1][0])
+        self.obstacles = np.where(self.grid == CELL.WALL)
+        self.obstacles = list(zip(self.obstacles[0], self.obstacles[1]))
+
+        self.gamma = gamma
+        self.num_episodes = num_episodes
+        self.train_epsilon = train_epsilon
+        self.test_epsilon = test_epsilon
+        
+        self.Q = defaultdict(lambda: defaultdict(float))
+        self.returns = defaultdict(lambda: defaultdict(list))
+
+        self.policy = Policy(train_epsilon, test_epsilon)
+        self.policy_filename = policy_filename
+        self.is_trained = False
+
+        if load_policy:
+            self.policy.load(policy_filename)
+            self.is_trained = True
+            self.policy.set_trained()
+
+    @staticmethod
+    def get_legal_actions(state):
+        candidates = MonteCarlo.actions.values()
+        legal_actions = []
+        for c in candidates:
+            combined = (state[2] + c[0], state[3] + c[1])
+            if np.all(np.abs(combined) <= MonteCarlo.max_velocity) and np.any(combined):
+                legal_actions.append(c)
+
+        legal_actions = [list(MonteCarlo.actions.keys())[list(MonteCarlo.actions.values()).index(a)] for a in legal_actions]
+        return legal_actions
+    
+    def generate_episode(self, start=None):
+        
+        episode = []
+        if start is None:
+            random_start_index = np.random.choice(len(self.starts))
+            random_start = self.starts[random_start_index]
+            start = random_start
+        state = start + (0, 0)
+        while state[:2] != self.goal:
+            action = self.policy[state]
+            dvx, dvy = MonteCarlo.actions[action]
+            vx, vy = state[2] + dvx, state[3] + dvy
+            # Ensure velocity stays within bounds
+            vx = max(-self.max_velocity, min(self.max_velocity, vx))
+            vy = max(-self.max_velocity, min(self.max_velocity, vy))
```

Comment:
i think we should only have the upper bound, but no lower bound

Reply:
I know that the task description allows your suggestion, however i am almost certain that this was a typo.